src/tunnel.rs

```diff
@@ -235,16 +269,60 @@ impl FactoryComponent for Tunnel {
         sender: relm4::FactorySender<Self>,
     ) {
         match msg {
+            // In the Toggle message handler:
             Self::Input::Toggle => {
-                match self.try_toggle() {
-                    Ok(_) => self.active = !self.active,
-                    Err(err) => sender
-                        .output_sender()
-                        .emit(Self::Output::Error(err.to_string())),
-                };
-                debug!("connection state: {}", self.active);
-                widgets.switch.set_state(self.active);
+                // Only validate when activating (not when deactivating)
+                if !self.active {
+                    if !self.saved {
+                        sender.output_sender().emit(TunnelOutput::Error(
+                            "You must save the configuration before activating the tunnel.".into(),
+                        ));
+                        widgets.switch.set_state(false);
+                        return;
+                    }
+
+                    if let Err(err) = self.is_cfg_valid() {
+                        sender
+                            .output_sender()
+                            .emit(TunnelOutput::Error(err.to_string()));
+                        widgets.switch.set_state(false);
+                        return;
+                    }
+                }
+                // Capture needed data before moving into async block
+                let tunnel_name = self.name.clone();
+                let tunnel_path = self.path();
+                let current_active = self.active;
+
+                sender.oneshot_command(async move {
```

Comment:
This should probably be `sender.spawn_oneshot_command(move || {`, otherwise the non-async methods will block whichever async runtime is running the request (or at least one execution thread that may not be able to properly handle it).